library/core/tests/num/int_sqrt.rs

```diff
@@ -0,0 +1,365 @@
+//! This tests the `Integer::isqrt` methods.
+
+macro_rules! tests {
+    ($($SignedT:ident $UnsignedT:ident),+) => {
+        $(
+            mod $SignedT {
+                /// This takes an input and, if it's nonnegative or
+                #[doc = concat!("`", stringify!($SignedT), "::MIN`,")]
+                /// checks that `isqrt` and `checked_isqrt` produce equivalent
+                /// results for that input and for the negative of that input.
+                fn isqrt_consistency_check(n: $SignedT) {
+                    // `$SignedT::MIN` will be negative, so we don't want to handle `n` as if it's nonnegative.
+                    if n >= 0 {
+                        assert_eq!(
+                            Some(n.isqrt()),
+                            n.checked_isqrt(),
+                            "`{n}.checked_isqrt()` should match `Some({n}.isqrt())`.",
+                        );
+                    }
+
+                    // `wrapping_neg` so that `$SignedT::MIN` will negate to
+                    // itself rather than panicking.
+                    let negative_n = n.wrapping_neg();
+
+                    // `negative_n` should be negative, but `n` could be zero,
+                    // so make sure not to check that one.
+                    if negative_n < 0 {
+                        assert_eq!(
+                            negative_n.checked_isqrt(),
+                            None,
+                            "`({negative_n}).checked_isqrt()` should be `None`, as {negative_n} is negative.",
+                        );
+
+                        ::std::panic::catch_unwind(::core::panic::AssertUnwindSafe(|| (-n).isqrt())).expect_err(
+                            &format!("`({negative_n}).isqrt()` should have panicked, as {negative_n} is negative.")
+                        );
+                    }
+                }
+
+                // Check that the following produce the correct values from
+                // `isqrt` and that `checked_isqrt` produces the same numeric
+                // value as `isqrt`. Check also that their negative versions
+                // and `$SignedT::MIN` produce a panic from `isqrt` and `None`
+                // from `checked_isqrt`:
+                //
+                // * the first and last 128 nonnegative values
+                // * powers of two, minus one
+                // * powers of two
+                #[test]
+                fn isqrt() {
```

Comment:
You could probably factor `isqrt` and `isqrt_extended` into their own macro, `tests_inner!($SignedT, simple_iter, near_max_iter, near_min_iter)`, that would work for both signed and unsigned. Then the only thing in the signed/unsigned modules is `isqrt_consistency_check` and the range calculations.

Reply:
OK, I've deduplicated the tests. Now the main macro takes the type name and a secondary macro name to generate the consistency check function. It produces a module that contains the consistency check and a version of both `isqrt` and `isqrt_extended` that works for both signed and unsigned types.